src/shared/lib/AnnotationUtils.ts

```diff
@@ -45,18 +50,113 @@ export function geneToMyCancerGenome(myCancerGenomes:IMyCancerGenome[]):IMyCance
 }
 
-export function geneAndProteinPosToHotspots(hotspots:HotspotMutation[]):{[key:string]: boolean}
+export function indexHotspots(hotspots:HotspotMutation[], defaultHotspotType?: string): IHotspotIndex
 {
-    // key: geneSymbol_proteinPosition
-    // (protienPosition: start[_end])
-    const map: {[key:string]: boolean} = {};
+    const index:IHotspotIndex = {};
 
-    // create a map for a faster lookup
-    hotspots.forEach(function(hotspot:HotspotMutation) {
-        const positions = hotspot.residue.match(/[0-9]+/g) || []; // start (and optionally end) positions
-        const key = [hotspot.hugoSymbol.toUpperCase()].concat(positions).join("_");
-        map[key] = true;
+    hotspots.forEach((hotspot:HotspotMutation) => {
+        let geneIndex = index[hotspot.hugoSymbol.toUpperCase()];
+
+        if (!geneIndex)
+        {
+            geneIndex = {};
+
+            index[hotspot.hugoSymbol.toUpperCase()] = geneIndex;
+        }
+
+        const hotspotType = hotspot.type || defaultHotspotType || "hotspot";
+        let typeIndex = index[hotspot.hugoSymbol.toUpperCase()][hotspotType];
+
+        if (!typeIndex)
+        {
+            typeIndex = {
+                hotspotMutations: [],
+                hotspotSet: new HotspotSet([]) // this is a placeholder to avoid TS errors.
+                                               // overwritten by the initHotspotSets function.
+            };
+
+            index[hotspot.hugoSymbol.toUpperCase()][hotspotType] = typeIndex;
+        }
+
+        typeIndex.hotspotMutations.push(hotspot);
     });
 
-    return map;
+    initHotspotSets(index);
+
+    return index;
+}
+
+export function initHotspotSets(hotspotIndex: IHotspotIndex)
+{
+    _.keys(hotspotIndex).forEach((geneSymbol:string) => {
+        _.keys(hotspotIndex[geneSymbol]).forEach((hotspotType:string) => {
+            const intervals = hotspotIndex[geneSymbol][hotspotType].hotspotMutations.map(
+                (hotspot:HotspotMutation):[number, number] => {
+                    // start (and optionally end) positions
+                    const positions = hotspot.residue.match(/[0-9]+/g) || [];
+
+                    let start = -1;
+                    let end = -1;
+
+                    if (positions.length > 0) {
+                        start = parseInt(positions[0]);
+                        end = start;
+                    }
+
+                    // overwrite end if exists
+                    if (positions.length > 1) {
+                        end = parseInt(positions[1]);
+                    }
+
+                    return [start, end];
+                }
+            );
+
+            hotspotIndex[geneSymbol][hotspotType].hotspotSet = new HotspotSet(intervals);
+        });
+    });
+}
+
+export function isHotspot(mutation:Mutation, index:IHotspotIndex):boolean
+{
+    let isHotspot = false;
+    let hotspotType:string|null = null;
+
+    if (mutation.mutationType.toLowerCase().indexOf("missense") > -1) {
```

Comment:
save `mutation.mutationType.toLowerCase()` in a variable instead of recomputing it a few times here